orttraining/orttraining/eager/opgen/opgen/generator.py

```diff
@@ -328,7 +328,44 @@ def _write_function_body(self, writer: writer.SourceWriter, mapped_func: MappedO
                 )
                 writer.writeline()
 
+        # Gather return info on the torch func, such as Tensor(a! -> a)
+        # which implies the return value is writeable tensor (Tensor&)
+        return_info = cpp_func.torch_func.return_type if cpp_func.torch_func else None
+
+        # if the torch func has a return ref tensor, out is the last param, and self param is the first input
+        # then we need to update and return out.
+        set_out_tensor = False
+        last_param = cpp_func.parameters[-1].member
+        if (
+            return_info
+            and last_param
+            and last_param.identifier.value == "out"
+            and first_param.identifier.value == "self"
+        ):
+
+            # output_alias is how the return tensor is marked, normally (a! -> a)
+            output_alias = self._get_alias_info(return_info)
+
+            for torch_p in last_param.torch_param:
+                # Confirm we have an output alias and that it is writable (!)
+                # and the current param (torch_p/last_param) is marked with the output alias
+                if output_alias and output_alias.is_writable and self._get_alias_info(torch_p) == output_alias:
+                    set_out_tensor = True
+                    writer.writeline("// resize the output and then create output ort value to be updated.")
+                    writer.writeline(
+                        "resize_output(invoker, dynamic_cast<ORTTensorImpl*>(out.unsafeGetTensorImpl()), self.sizes());"
+                    )
+                    writer.writeline("auto ort_input_out = create_ort_value(invoker, out);")
+                    writer.writeline()
+
+        cast_output_to_bool = False
```

Comment:
this is named wrong. It should be cast_output_to_out_type. We are not casting to bool but are respecting the type in the output variable.

Reply:
removed entirely now.